### PR TITLE
KAFKA-10816: Add health check endpoint for Kafka Connect

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-10816), [KIP-1017](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1017%3A+Health+check+endpoint+for+Kafka+Connect)

Adds the `GET /health` endpoint to the Kafka Connect REST API, and covers all three states (starting, healthy, unhealthy) in both modes (distributed, standalone) with integration tests. Some unit tests are also added.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@gharris1727, if you have time would you mind taking a look? I know the diff is fairly large but most of that is either small-but-far-reaching changes (i.e., getting rid of the `DistributedHerder::running` field and replacing its uses in our integration tests with the health check endpoint) or newly-added unit and integration tests. Still, if this is too much to do in one round, please let me know and I can try to break it out into smaller PRs.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: nit: have a single instance that is always returned? I like the method, rather than having a final constant.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: These are pretty natural methods, and I like them. And even though these are not part of the public api...

I think it is a bit odd that these rest entities would be aware of an internal abstraction `Stage`. It's not the first time, as TaskInfo depends on ConnectorTaskId, ServerInfo on AppInfoParser, etc.

Just a thought that crossed my mind, but I don't have an alternative that I like better so this can probably stay as-is.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:129]: Please call `Thread#join()` at some point to ensure the thread gets cleaned up :)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/RootResource.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/RootResource.java:94]: Should this be wrapped in a ConnectRestException, so it gets picked up by the ConnectExceptionMapper? It falls back to a 500 otherwise so it doesn't have a lot of difference, and it should never throw an ExecutionException, but 🤷

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnect.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnect.java:141]: Does the order here matter? An exception starting httpClient still leaves the kafkaCluster started.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:245]: Was there a problem where the log directories would be deleted before the brokers shut down? Nice find!

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/WorkerHandle.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/WorkerHandle.java:71]: This appears to have been unused. I can't imagine a situation where you'd want to programmatically search for the name, so it should probably only be visible in the toString() :+1:

**Review by gharris1727 — COMMENTED** [review]
Thanks for the feature, Chris!

It mostly looks good to me. I had a few minor comments i'd like your thoughts on.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnect.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnect.java:141]: We need to start the HTTP client before we start the Connect cluster in order to be able to hit the cluster's REST API before `startConnect` has returned (which is only possible in the standalone integration test because we run `EmbeddedConnectStandalone::start` on a separate thread from the one we use to make REST requests).

This is dirty-bordering-on-filthy, but I couldn't think of a better way to probe a worker's health check endpoint during simulated blocked startup. Thoughts?

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:245]: Nah, this was just because there was a race in one of our integration tests where we shut down the underlying Kafka cluster and then wait for a fixed time period before asserting that the Connect cluster is unavailable. Making Kafka cluster shutdown synchronous fixed that problem.

(And the method decomposition was just to appease Checkstyle, although I do have to admit I also find it easier to read now.)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: It should be easy enough to change the parameter to something like `String statusDetail` and leave it to the caller to invoke `Stage::summarize` if they're using the stage class to track progress. Did that not seem better to you? (I'd be fine with it.)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/RootResource.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/RootResource.java:94]: The exception mapper will catch this anyways, won't it? The intent was to hit [this part](https://github.com/apache/kafka/blob/d0dfefbe6394276eb329b6ca998842a984add506/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/errors/ConnectExceptionMapper.java#L73-L77) of the code path and generate the standard 500-on-unexpected-error response.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:245]: I gave this a shot, let me know what you think.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:129]: Ah, good call. Done 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: I gave this a shot, let me know what you think.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: 👍 done

**Comment by C0urante** [comment]
Thanks for the review @gharris1727! This is ready for another pass when you have time.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java]: Ah yeah this is very reasonable. For some reason I thought there were a lot of call-sites that would need an extra null guard, but that wasn't the case. This is great.

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks @C0urante!

**Comment by C0urante** [comment]
Thanks Greg!

I've realized the TODO in the code base was unnecessary and this should be safe to merge without waiting for more green CI runs (see the latest commit message for more details).

Merging...

**Pull request merged**
